File: app/src/main/java/com/example/findmeinlol/viewmodel/adapter/SearchViewRecyclerAdapter.java
package com.example.findmeinlol.viewmodel.adapter;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.findmeinlol.databinding.RecyclerviewItemBinding;
import com.example.findmeinlol.model.data.SummonerDto;
import com.example.findmeinlol.model.SearchModel;
import com.example.findmeinlol.viewmodel.SearchViewModel;

public class SearchViewRecyclerAdapter
        extends RecyclerView.Adapter<SearchViewRecyclerAdapter.ViewHolder> {

    SearchViewModel mSearchViewModel;
    SearchModel mSearchModel;
    CallBackListener mCallBackListener;

    public SearchViewRecyclerAdapter(SearchViewModel searchViewModel) {
        this.mSearchViewModel = searchViewModel;
        mSearchModel = mSearchViewModel.getSearchModel();
    }

    public void setCallBackListener(CallBackListener mCallBackListener) {
        this.mCallBackListener = mCallBackListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                                                                   int viewType) {
        RecyclerviewItemBinding binding = RecyclerviewItemBinding.inflate(
                LayoutInflater.from(parent.getContext()),
                parent,
                false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        SummonerDto summonerDto = mSearchViewModel.getSearchModel().getSummoner(position);
        holder.recyclerviewItemBinding.setSummonerDto(summonerDto);
        holder.recyclerviewItemBinding.recyclerviewItemImgUser.setImageBitmap(summonerDto.getProfileIcon());
        holder.recyclerviewItemBinding.recyclerviewItemImgTier.setImageURI(Uri.parse(summonerDto.getTierIconId()));
        setListener(holder, position);
    }

    private void setListener(ViewHolder holder, int position) {
        holder.recyclerviewItemBinding.recyclerviewItemBtnFavorite.setOnClickListener(v -> {
            mCallBackListener.favoriteBtnClicked(position,
                    holder.recyclerviewItemBinding.recyclerviewItemBtnFavorite.isChecked());
        });

        holder.recyclerviewItemBinding.recyclerviewItemBtnDelete.setOnClickListener(v -> {
            mCallBackListener.deleteBtnClicked(position);
        });

        holder.itemView.setOnClickListener(v -> {
            mCallBackListener.itemClicked(position);
        });
    }

    @Override
    public int getItemCount() {
        return mSearchModel.getSize();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private RecyclerviewItemBinding recyclerviewItemBinding;

        ViewHolder(RecyclerviewItemBinding recyclerviewItemBinding) {
            super(recyclerviewItemBinding.getRoot());
            this.recyclerviewItemBinding = recyclerviewItemBinding;
        }
    }
}

File: app/src/main/java/com/example/findmeinlol/viewmodel/adapter/SearchViewRecyclerAdapter.kt
package com.example.findmeinlol.viewmodel.adapter

import android.net.Uri
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.findmeinlol.databinding.RecyclerviewItemBinding
import com.example.findmeinlol.model.data.Summoner
import com.example.findmeinlol.view.CallBack
import com.example.findmeinlol.viewmodel.SearchViewModel

class SearchViewRecyclerAdapter(searchViewModel: SearchViewModel):
    RecyclerView.Adapter<SearchViewRecyclerAdapter.ViewHolder> () {
    private val searchViewModel = searchViewModel
    private lateinit var callback: CallBack

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): SearchViewRecyclerAdapter.ViewHolder {
        val binding: RecyclerviewItemBinding = RecyclerviewItemBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: SearchViewRecyclerAdapter.ViewHolder, position: Int) {
        val summoner: Summoner = searchViewModel.getSummoner(position)
        holder.recyclerviewItemBinding.summoner = summoner
        holder.recyclerviewItemBinding.recyclerviewItemImgUser.setImageBitmap(summoner.profileIcon)
        holder.recyclerviewItemBinding.recyclerviewItemImgTier.setImageURI(Uri.parse(summoner.tierIconId))
        setListener(holder, position)
    }

    override fun getItemCount(): Int {
        return searchViewModel.getSearchModel().getSize()
    }

    private fun setListener(holder: ViewHolder, position: Int) {
        holder.recyclerviewItemBinding.recyclerviewItemBtnDelete.setOnClickListener{
            searchViewModel.deleteSummoner(position)
        }
        holder.itemView.setOnClickListener{
            callback.onItemClicked(position)
        }
    }

    fun setListener(c: CallBack) {
        this.callback = c
    }

    inner class ViewHolder(binding: RecyclerviewItemBinding) : RecyclerView.ViewHolder(binding.root) {
        val recyclerviewItemBinding = binding
    }
}